presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java

```diff
@@ -149,10 +150,8 @@
                 newSourceBuilder.build(),
                 inputsBuilder.build());
 
-        if (!result.getOutputSymbols().equals(project.getOutputSymbols())) {
-            // we need to strip unnecessary symbols (hash, partitioning columns).
-            result = new ProjectNode(idAllocator.getNextId(), result, identity(project.getOutputSymbols()));
-        }
+        // we need to strip unnecessary symbols (hash, partitioning columns).
+        result = restrictOutputs(idAllocator, result, ImmutableSet.copyOf(project.getOutputSymbols()));
```

Comment:
What if `project` had a larger set of output symbols than `result` node? Is such situation legal? Should we add some assertion here? Because `restrictOutputs` clearly allows node having only a subset of `permittedOutputs`, which is not what was checked here before (sets had to be equal).
I'm leaning towards assertion here.